File: src/lib/format.ts
import { format, formatDistanceToNowStrict, isToday, isTomorrow } from 'date-fns'

export const day = (d: string | Date) => format(new Date(d), 'd MMM yyyy')
export const dayTime = (d: string | Date) => format(new Date(d), 'd MMM yyyy, h:mm a')
export const time = (d: string | Date) => format(new Date(d), 'h:mm a')
export const ago = (d: string | Date) => `${formatDistanceToNowStrict(new Date(d))} ago`

export function relativeDay(d: string | Date) {
  const date = new Date(d)
  if (isToday(date)) return 'Today'
  if (isTomorrow(date)) return 'Tomorrow'
  return format(date, 'EEE d MMM')
}

/** enum_value → Enum Value */
export const humanise = (s: string | null | undefined) =>
  (s ?? '').replace(/_/g, ' ').replace(/\b\w/g, (c) => c.toUpperCase())

export const DESIGNATION: Record<string, string> = {
  senior_advocate: 'Senior Advocate',
  advocate: 'Advocate',
  advocate_on_record: 'Advocate-on-Record',
}

export const CATEGORY: Record<string, string> = {
  general: 'General',
  court_notice: 'Court Notice',
  condolence: 'Condolence',
  election: 'Election',
  welfare_scheme: 'Welfare Scheme',
  meeting_notice: 'Meeting Notice',
  urgent: 'Urgent',
}

export const ENTRY_TYPE: Record<string, string> = {
  court_holiday: 'Court Holiday',
  association_meeting: 'Association Meeting',
  gbm_egm: 'GBM / EGM',
  event: 'Event',
  election: 'Election',
  hearing_of_interest: 'Hearing of Interest',
  other: 'Other',
}

/** Calendar colour-coding — holidays must read differently at a glance (PRD 3.4). */
export const ENTRY_COLOUR: Record<string, string> = {
  court_holiday: 'bg-rule-wash text-rule border-rule-soft',
  association_meeting: 'bg-ink-100 text-ink-800 border-ink-200',
  gbm_egm: 'bg-amber-100 text-amber-900 border-amber-200',
  event: 'bg-emerald-100 text-emerald-900 border-emerald-200',
  election: 'bg-violet-100 text-violet-900 border-violet-200',
  hearing_of_interest: 'bg-sky-100 text-sky-900 border-sky-200',
  other: 'bg-paper-edge text-ink-700 border-paper-edge',
}

export const EVENT_TYPE: Record<string, string> = {
  seminar: 'Seminar',
  cle_training: 'CLE / Training',
  cultural: 'Cultural',
  sports: 'Sports',
  felicitation: 'Felicitation',
  agm: 'AGM',
  farewell: 'Farewell',
  other: 'Other',
}

export const TICKET_CATEGORY: Record<string, string> = {
  general: 'General Enquiry',
  membership: 'Membership',
  welfare_scheme: 'Welfare Scheme',
  grievance: 'Grievance',
  technical_support: 'Technical Support',
  other: 'Other',
}

export const VISIBILITY: Record<string, string> = {
  all_members: 'All Members',
  committee_only: 'Committee Only',
  office_bearers_only: 'Office Bearers Only',
}

export const fileSize = (b?: number | null) =>
  !b ? '—' : b > 1e6 ? `${(b / 1e6).toFixed(1)} MB` : `${Math.round(b / 1e3)} KB`

File: mobile/lib/format.dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import 'theme.dart';

/* Ported from src/lib/format.ts so the app labels a record exactly as the web does. */

DateTime at(dynamic v) => DateTime.parse(v.toString()).toLocal();

String day(dynamic d) => DateFormat('d MMM yyyy').format(at(d));
String dayTime(dynamic d) => DateFormat('d MMM yyyy, h:mm a').format(at(d));
String time(dynamic d) => DateFormat('h:mm a').format(at(d));
String monthYear(DateTime d) => DateFormat('MMMM yyyy').format(d);
String monthAbbrev(DateTime d) => DateFormat('MMM').format(d);

/// "Tuesday, 4 August 2026" — the dashboard's dateline, as the web sets it
/// with toLocaleDateString('en-IN', { weekday, day, month, year }).
String dayOfWeekLine(DateTime d) => DateFormat('EEEE, d MMMM yyyy').format(d);

/// "3 days ago" — the strict, single-unit form date-fns produces.
String ago(dynamic d) {
  final diff = DateTime.now().difference(at(d));
  final s = diff.inSeconds.abs();
  final (n, unit) = switch (s) {
    < 60 => (s, 'second'),
    < 3600 => (s ~/ 60, 'minute'),
    < 86400 => (s ~/ 3600, 'hour'),
    < 2592000 => (s ~/ 86400, 'day'),
    < 31536000 => (s ~/ 2592000, 'month'),
    _ => (s ~/ 31536000, 'year'),
  };
  return '$n $unit${n == 1 ? '' : 's'} ago';
}

bool _sameDay(DateTime a, DateTime b) =>
    a.year == b.year && a.month == b.month && a.day == b.day;

String relativeDay(dynamic d) {
  final date = at(d);
  final now = DateTime.now();
  if (_sameDay(date, now)) return 'Today';
  if (_sameDay(date, now.add(const Duration(days: 1)))) return 'Tomorrow';
  return DateFormat('EEE d MMM').format(date);
}

/// enum_value → Enum Value
String humanise(String? s) => (s ?? '')
    .split('_')
    .where((w) => w.isNotEmpty)
    .map((w) => w[0].toUpperCase() + w.substring(1))
    .join(' ');

const designationLabel = {
  'senior_advocate': 'Senior Advocate',
  'advocate': 'Advocate',
  'advocate_on_record': 'Advocate-on-Record',
};

const categoryLabel = {
  'general': 'General',
  'court_notice': 'Court Notice',
  'condolence': 'Condolence',
  'election': 'Election',
  'welfare_scheme': 'Welfare Scheme',
  'meeting_notice': 'Meeting Notice',
  'urgent': 'Urgent',
};

const entryTypeLabel = {
  'court_holiday': 'Court Holiday',
  'association_meeting': 'Association Meeting',
  'gbm_egm': 'GBM / EGM',
  'event': 'Event',
  'election': 'Election',
  'hearing_of_interest': 'Hearing of Interest',
  'other': 'Other',
};

const eventTypeLabel = {
  'seminar': 'Seminar',
  'cle_training': 'CLE / Training',
  'cultural': 'Cultural',
  'sports': 'Sports',
  'felicitation': 'Felicitation',
  'agm': 'AGM',
  'farewell': 'Farewell',
  'other': 'Other',
};

const ticketCategoryLabel = {
  'general': 'General Enquiry',
  'membership': 'Membership',
  'welfare_scheme': 'Welfare Scheme',
  'grievance': 'Grievance',
  'technical_support': 'Technical Support',
  'other': 'Other',
};

const visibilityLabel = {
  'all_members': 'All Members',
  'committee_only': 'Committee Only',
  'office_bearers_only': 'Office Bearers Only',
};

String fileSize(dynamic b) {
  final n = b is num ? b.toInt() : int.tryParse('${b ?? ''}') ?? 0;
  if (n <= 0) return '—';
  if (n > 1000000) return '${(n / 1000000).toStringAsFixed(1)} MB';
  return '${(n / 1000).round()} KB';
}

/// Calendar colour-coding — holidays must read differently at a glance (PRD 3.4).
/// The palette is monochrome, so the types separate by treatment instead of hue:
/// red fill, solid black, outline, grey fill, hairline.
({Color bg, Color fg, Color border}) entryStyle(String type) => switch (type) {
  'court_holiday' => (bg: C.accent, fg: C.onDark, border: C.accent),
  'gbm_egm' => (bg: C.ink, fg: C.onDark, border: C.ink),
  'election' => (bg: C.accentWash, fg: C.accent, border: C.accentWash),
  'association_meeting' => (bg: C.surface, fg: C.ink, border: C.ink),
  'event' => (bg: C.sunk, fg: C.ink2, border: C.sunk),
  'hearing_of_interest' => (bg: C.surface, fg: C.ink3, border: C.separator),
  _ => (bg: C.canvas, fg: C.ink4, border: C.separator),
};
